File: Client/Chuangshi/Skill/Effect/EffectCreateEntity.cpp
#include "EffectCreateEntity.h"
#include "Util/ConvertUtil.h"
#include "Util/GolbalBPFunctionLibrary.h"
#include "Interface/Monster/CSkillInterface.h"
#include "Skill/SkillBase/Skill.h"

UEffectCreateEntity::UEffectCreateEntity()
{
}

UEffectCreateEntity::~UEffectCreateEntity()
{
}

void UEffectCreateEntity::init(FHIT_EFFECT dictDat, USkill * skill)
{
	UEffectSkillBase::init(dictDat, skill);
	_scriptID = dictDat.Param1;
	if (dictDat.Param2 != "")
	{
		TArray<FString> Param2 = UGolbalBPFunctionLibrary::SplitString(dictDat.Param1, false, TEXT("|"));
		_offsetFront = FSTRING_TO_FLOAT(Param2[0]);
		if (Param2.Num() >= 2)
		{
			_offsetLeft = FSTRING_TO_FLOAT(Param2[1]);
		}
	}
}

void UEffectCreateEntity::onReceive(USkill * skill, CS3Entity * caster, CS3Entity * receiver)
{
	float yaw = receiver->Direction().Z * PI / 180;
	FVector direction = receiver->Direction();
	FVector position = receiver->Position();

	if (_offsetFront != 0.0)
	{
		position.Y += _offsetFront * FMath::Sin(yaw);
		position.X += _offsetFront * FMath::Cos(yaw);
	}
	if (_offsetLeft != 0.0)
	{
		position.Y += _offsetLeft * FMath::Sin(yaw - PI / 2);
		position.X += _offsetLeft * FMath::Cos(yaw - PI / 2);
	}

	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), receiver->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->CSkillCreateEntityByScriptID(_scriptID, position, direction);
	}
}

File: Server/scripts/cell/ConfigObject/Skill/Effect/EffectCreateEntity.py
# -*- coding: utf-8 -*-

from ConfigObject.Skill.Effect.EffectSkillBase import EffectSkillBase
import KBEDebug
import Math
import math
import KST
import csdefine

class EffectCreateEntity( EffectSkillBase ):
	"""
	目标位置创建entity
	"""
	def __init__( self ):
		EffectSkillBase.__init__( self )
		self._scriptID = ""
		self._offsetFront = 0.0
		self._offsetLeft = 0.0
	
	def init( self, dictDat, skill ):
		"""
		virtual method;
		读取技能配置
		@param dictDat: 配置数据 { "Param1":"", "Param2":"", "Param3":"" }
		@type  dictDat: python dict
		"""
		EffectSkillBase.init( self, dictDat, skill )
		self._scriptID = dictDat["Param1"]
		
		if dictDat["Param2"]:
			offset = dictDat["Param2"].split("|")
			self._offsetFront = float( offset[0] )
			if len(offset) >= 2:
				self._offsetLeft = float( offset[1] )

	def onReceive( self, skill, caster, receiver ):
		"""
		效果开始
		"""
		yaw = receiver.direction.z
		position = Math.Vector3( receiver.position )
		if self._offsetFront:
			position.x += self._offsetFront * math.sin( yaw )
			position.z += self._offsetFront * math.cos( yaw )
		if self._offsetLeft:
			position.x += self._offsetLeft * math.sin( yaw - math.pi/2 )
			position.z += self._offsetLeft * math.cos( yaw - math.pi/2 )
			
		object_ = KST.g_objFactory.getObject( self._scriptID )
		if(object_):
			params = {}
			if not object_.hasScriptFlag( csdefine.FLAG_CAN_FLY ):
				rayCastList = receiver.navmeshRaycast( receiver.position, position )
				if rayCastList:
					position = rayCastList[0]
					params = { "parent" : receiver.parent }
			params.update( {"spawnPos": position, "spawnDir":Math.Vector3( receiver.direction ) } )
			receiver.createEntityByScriptID( self._scriptID, position, Math.Vector3( receiver.direction ), params )



